docs/content/sips/00x-cloudevent-trigger.md

```diff
@@ -0,0 +1,263 @@
+
+
+## Proposal
+
+This document proposes adding features in the spin SDK to support CloudEvents. CloudEvents itself is a envelop for the underlying transport protocol, such as AMQP, Kafka, HTTP, etc. This proposal aims at providing a CloudEvent component for the [HTTP Protocol Bindings](https://github.com/cloudevents/spec/blob/main/cloudevents/bindings/http-protocol-binding.md). Here is an example shows the mapping of an event with an HTTP POST request in CloudEvent's binary format.
+```
+POST /someresource HTTP/1.1
+Host: webhook.example.com
+ce-specversion: 1.0
+ce-type: com.example.someevent
+ce-time: 2018-04-05T03:56:24Z
+ce-id: 1234-1234-1234
+ce-source: /mycontext/subcontext
+    .... further attributes ...
+Content-Type: application/json; charset=utf-8
+Content-Length: nnnn
+{
+    ... application data ...
+}
+```
+
+Creating an HTTP CloudEvents trigger is done by defining the top level application trigger in spin. The following code snippet shows the definition of a HTTP CloudEvents trigger.
+```toml
+# spin.toml
+trigger = { type = "http", base = "/", schema = "cloudevents" }
+```
+
+The added `schema` attribute in trigger will tell spin application that it will expect the HTTP request and responses are CloudEvents.
+
+> Note that the `schema` attribute is not the same as the `schema` attribute in the CloudEvents spec. The `schema` attribute in the spec is the schema of the payload.
+
+We also allow users to define individual component to be CloudEvents components. For example, we could define a HTTP CloudEvents component in spin.toml:
+
+```toml
+# spin.toml
+trigger = { type = "http", base = "/" }
+
+[[component]]
+id = "hello"
+source = "target/wasm32-wasi/release/spinhelloworld.wasm"
+description = "A simple component that returns hello."
+schema = "cloudevents"
+[component.trigger]
+route = "/hello"
+```
+
+> Note that if there is no `schema` attribute in the application or individual component, the HTTP request and responses are not CloudEvents.
+
+## The WebAssembly interface
+
+The CloudEvent trigger is built on top of the
+[WebAssembly component model](https://github.com/WebAssembly/component-model).
+The current interface is defined using the
+[WebAssembly Interface (WIT)](https://github.com/bytecodealliance/wit-bindgen/blob/main/WIT.md)
+format, and is a function that takes the event payload as its only parameter:
+
+```fsharp
+// wit/ephemeral/spin-ce.wit
+
+// The event payload.
+record event {
+    // The event type.
+    type: string,
+    // The event id.
+    id: string,
+    // The event source.
+    source: string,
+    // The event specversion.
+    specversion: string,
+    // The event data content type.
+    datacontenttype: string,
+    // The event data schema.
+    dataschema: string,
+    // The event subject.
+    subject: string,
+    // The event time.
+    time: option<string>,
+    // The event data.
+    data: option<string>
```

Comment:
Should this be `option<list<u8>>` instead?

Reply:
yes